File: LeetCode/Top Interview 150/[240723] 238. Product of Array Except Self/정훈.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prd = 1
        zcnt = 0
        for n in nums:
            if n==0:
                zcnt += 1
                continue
            prd *= n
        if zcnt > 1:
            return [0 for _ in range(len(nums))]
        if zcnt > 0:
            return [ 0 if n != 0 else prd for n in nums]
        answer=[]
        for n in nums:
            answer.append(int(prd/n))
        return answer
# 나누기 쓰면 안됨

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        cur = 1
        ans = [1 for _ in range(len(nums))]
        for i in range(len(nums)):
            ans[i] *= cur
            cur *= nums[i]
        cur = 1
        for i in range(len(nums)-1, -1, -1):
            ans[i] *= cur
            cur *= nums[i]
        return ans

File: LeetCode/Top Interview 150/[240723] 238. Product of Array Except Self/정훈.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int ans[] = new int[n];
        Arrays.fill(ans, 1);
        // 앞 뒤로 누적곱 구하기

        // e.g. 1 2 3
        int cur = 1;
        for (int i = 0; i < n; i++) {
            ans[i] *= cur;
            cur *= nums[i];
        }
        // 1 1 1
        // 1 1 1
        // 1 1 1
        // 1 1 2
        cur = 1;
        for (int i = n-1; i>=0; i--) {
            ans[i] *= cur;
            cur *= nums[i];
        }
        // 1 1 2
        // 1 1 2
        // 1 3 2
        // 6 3 2
        return ans;
    }
}
